{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "40c5ca67",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import textwrap\n",
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "from tqdm import tqdm\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "261da170",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1875 entries, 0 to 1874\n",
      "Data columns (total 28 columns):\n",
      " #   Column                      Non-Null Count  Dtype  \n",
      "---  ------                      --------------  -----  \n",
      " 0   pubmed_id                   1841 non-null   float64\n",
      " 1   title                       1841 non-null   object \n",
      " 2   abstract                    1839 non-null   object \n",
      " 3   date                        1841 non-null   object \n",
      " 4   author                      1841 non-null   object \n",
      " 5   journal                     1841 non-null   object \n",
      " 6   name                        1875 non-null   object \n",
      " 7   aaseq                       1852 non-null   object \n",
      " 8   comment                     1175 non-null   object \n",
      " 9   uniprot_id                  1788 non-null   object \n",
      " 10  function                    640 non-null    object \n",
      " 11  domain                      532 non-null    object \n",
      " 12  ph                          1804 non-null   float64\n",
      " 13  temperature                 1708 non-null   object \n",
      " 14  dilution                    1875 non-null   object \n",
      " 15  dialysis                    1875 non-null   object \n",
      " 16  column_filtration           1875 non-null   object \n",
      " 17  column_binding              1875 non-null   object \n",
      " 18  high_pressure               1875 non-null   object \n",
      " 19  other_method                1875 non-null   object \n",
      " 20  activity                    1875 non-null   object \n",
      " 21  solubility                  1875 non-null   object \n",
      " 22  non_aggregability           1875 non-null   object \n",
      " 23  circular_dichroism          1875 non-null   object \n",
      " 24  fluorescence_tryptophan     1875 non-null   object \n",
      " 25  nuclear_magnetic_resonance  1875 non-null   object \n",
      " 26  crystallization             1875 non-null   object \n",
      " 27  structure_determination     1875 non-null   object \n",
      "dtypes: float64(2), object(26)\n",
      "memory usage: 410.3+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "# Load the downloaded DB as a pandas DF\n",
    "\n",
    "df = pd.read_table(\"REFOLD_161028.txt\")\n",
    "print(df.info())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d165ac59",
   "metadata": {},
   "source": [
    "It looks like the newer DB is missing some important fields such as the solvant used in the experiments. For the older records (before 2009), the information could be scrapped from https://pford.info/refolddatabase/. But for the newer records, it might be necessary to explore the papers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "b8d82684",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**pubmed_id**: 11829501.0\n",
      "********************************************************************************\n",
      "**title**: Characterization of two highly similar Rad51 homologs of Physcomitrella patens.\n",
      "********************************************************************************\n",
      "**abstract**: The moss Physcomitrella patens, which is a land plant with efficient homologous\n",
      "recombination, encodes two Rad51 proteins (PpaRad51.1 and PpaRad51.2). The\n",
      "PpaRad51.1 and PpaRad51.2 proteins, which share 94  0dentity between them,\n",
      "interact with themselves and with each other. Both proteins bind ssDNA and dsDNA\n",
      "in a Mg(2+) and pH-dependent manner, with a stoichiometry of one PpaRad51.1\n",
      "monomer per 3(+/-1) nt or bp and one PpaRad51.2 monomer per 1(+/-0.5) nt or bp,\n",
      "respectively. At neutral pH, a 1.6-fold excess of both proteins is required for\n",
      "ssDNA and dsDNA binding. PpaRad51.1 and PpaRad51.2 show ssDNA-dependent ATPase\n",
      "activity and efficiently promote strand annealing in a nucleotide-independent\n",
      "but in a Mg(2+)-dependent manner. Both proteins promote joint-molecule\n",
      "formation, DNA strand invasion and are able to catalyse strand exchange in the\n",
      "presence of Mg(2+) and ATP. No further increase in the activities is observed\n",
      "when both proteins are present in the same reaction. None of the PpaRad51 gene\n",
      "products complement the DNA repair and recombination phenotype of Saccharomyces\n",
      "cerevisiae rad51delta mutants. However, PpaRad51.1 confers a dominant-negative\n",
      "DNA repair phenotype, and both PpaRad51 proteins reduce the levels of double-\n",
      "strand break-induced recombination when overexpressed in S. cerevisiae wt cells.\n",
      "These results suggest that both PpaRad51 proteins are bona fide Rad51 proteins\n",
      "that may contribute, in a different manner, to homologous recombination, and\n",
      "that they might replace ScRad51 in a hypothetical yeast protein complex\n",
      "inactivating different functions required for recombinational repair.\n",
      "********************************************************************************\n",
      "**date**: 2002 Feb\n",
      "********************************************************************************\n",
      "**author**: Ayora Silvia,Piruat José I,Luna Rosa,Reiss Bernd,Russo Vincenzo E A,Aguilera\n",
      "Andrés,Alonso Juan C\n",
      "********************************************************************************\n",
      "**journal**: Journal of molecular biology. 316(1) 35-49\n",
      "********************************************************************************\n",
      "**name**: Rad51A protein\n",
      "********************************************************************************\n",
      "**aaseq**: MATVSAATEGVTEEEVHHGPYLVEQLQNCGISSLDLKKLKDAGFCTVESVAYSAKKDLVLIKGLSDAKDKIIEAATKLVP\n",
      "MGFTSAKQMHEQRAELIQITTGAKEFDILEGGIETGSITEIYGEFRSEKSQICHTLCVTCQLPLDQGGGEGKALYIDAEG\n",
      "TFRPQRLLQIAEKYGLNGQDVLDNVAYARAYNTDHQMKLLVEAASMMAETRFALMVVDSSTALYRTDYSGRGELAARQVH\n",
      "LAKFLRGCQKIADEFGIAVVITNQVVAQVDGSAMFNGPQHKPIGGNIIAHASTTRLSVRKGRGEERVIKVVASPCLAEQE\n",
      "QRFQITNEGAVDVKE\n",
      "********************************************************************************\n",
      "**comment**: Sequence provided refers to recA protein only.\n",
      "********************************************************************************\n",
      "**uniprot_id**: Q93V52\n",
      "********************************************************************************\n",
      "**function**: nan\n",
      "********************************************************************************\n",
      "**domain**: nan\n",
      "********************************************************************************\n",
      "**ph**: 7.5\n",
      "********************************************************************************\n",
      "**temperature**: 25\n",
      "********************************************************************************\n",
      "**dilution**: f\n",
      "********************************************************************************\n",
      "**dialysis**: t\n",
      "********************************************************************************\n",
      "**column_filtration**: f\n",
      "********************************************************************************\n",
      "**column_binding**: f\n",
      "********************************************************************************\n",
      "**high_pressure**: f\n",
      "********************************************************************************\n",
      "**other_method**: f\n",
      "********************************************************************************\n",
      "**activity**: t\n",
      "********************************************************************************\n",
      "**solubility**: f\n",
      "********************************************************************************\n",
      "**non_aggregability**: f\n",
      "********************************************************************************\n",
      "**circular_dichroism**: f\n",
      "********************************************************************************\n",
      "**fluorescence_tryptophan**: f\n",
      "********************************************************************************\n",
      "**nuclear_magnetic_resonance**: f\n",
      "********************************************************************************\n",
      "**crystallization**: f\n",
      "********************************************************************************\n",
      "**structure_determination**: f\n",
      "********************************************************************************\n"
     ]
    }
   ],
   "source": [
    "# Extract a single record and explore it to see how it looks like\n",
    "\n",
    "record = df.sample(1).iloc[0]\n",
    "\n",
    "for col, val in record.items():\n",
    "    val_str = str(val)\n",
    "    wrapped = textwrap.fill(val_str, width=80)  # wrap at 80 chars\n",
    "    print(f\"**{col}**: {wrapped}\")\n",
    "    print(\"*\"*80)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2415ada",
   "metadata": {},
   "source": [
    "As the records don't even contains the link to to Monash University REFOLD database, we will have to match them ourselves. First we scrap Monash University REFOLD database to get the list of records.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "87b25ee9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# First we gather the columns name for our DB using a single record\n",
    "r = requests.get(\"https://pford.info/refolddatabase/refoldingrecord/5/\")\n",
    "\n",
    "soup = BeautifulSoup(r.text, \"html.parser\")\n",
    "header = \"\"\n",
    "col_names = []\n",
    "data = []\n",
    "\n",
    "for th in soup.find_all(\"th\"):\n",
    "    if th.get(\"class\")[0] == \"detail_header\":\n",
    "        header = th.find(string=True, recursive=False).strip()\n",
    "    else:\n",
    "        col_names.append(header + \".\" + th.text.replace(\" \",\"_\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "5251aa0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gatherrecord(url):\n",
    "    r = requests.get(url)\n",
    "    soup = BeautifulSoup(r.text, \"html.parser\")\n",
    "\n",
    "    data = []\n",
    "    for row in soup.find_all('tr'):\n",
    "            for cell in row.find_all(\"td\"):\n",
    "                data.append(cell.text)\n",
    "    \n",
    "    return data\n",
    "\n",
    "\n",
    "def analyze_page(url):\n",
    "    r = requests.get(url)\n",
    "    soup = BeautifulSoup(r.text, \"html.parser\")\n",
    "    page_data = []\n",
    "    for row in tqdm(soup.find_all('tr'), unit=\"record\"):\n",
    "        link = row.find(\"a\")\n",
    "        if link: # Deal with the headers \n",
    "            url = \"https://pford.info/refolddatabase/refoldingrecord/\" + link.get(\"href\")\n",
    "            row_data = gatherrecord(url)\n",
    "            page_data.append(row_data)\n",
    "                \n",
    "\n",
    "    return page_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "f596e4d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 26/26 [00:40<00:00,  1.58s/record]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "It took 44.68540167808533 to deal with the page\n",
      "Dealing with : ?page=2/47\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 26/26 [00:41<00:00,  1.58s/record]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "It took 44.633166790008545 to deal with the page\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Next we explore the full website\n",
    "\n",
    "full_data = []\n",
    "\n",
    "# We start from the start page and do it manually (as we need to have a next page)\n",
    "start_time = time.time()\n",
    "url = \"https://pford.info/refolddatabase/refoldingrecord\"\n",
    "r = requests.get(url)\n",
    "soup = BeautifulSoup(r.text, \"html.parser\")\n",
    "page_data = analyze_page(url)\n",
    "    \n",
    "full_data.extend(page_data)\n",
    "end_time = time.time()\n",
    "elapsed_time = end_time - start_time\n",
    "print(\"It took \"+ str(elapsed_time) + \" to deal with the page\")\n",
    "\n",
    "\n",
    "# We loop while we have a next page\n",
    "while soup.find(\"a\", string=\"Next\"):\n",
    "    start_time = time.time()\n",
    "\n",
    "    url = \"https://pford.info/refolddatabase/refoldingrecord\"+soup.find(\"a\", string=\"Next\").get(\"href\")\n",
    "    print(\"Dealing with : \"+soup.find(\"a\", string=\"Next\").get(\"href\") + \"/47\")\n",
    "    r = requests.get(url)\n",
    "    soup = BeautifulSoup(r.text, \"html.parser\")\n",
    "    page_data = analyze_page(url)\n",
    "    full_data.extend(page_data)\n",
    "\n",
    "    end_time = time.time()\n",
    "    elapsed_time = end_time - start_time\n",
    "    print(\"It took \"+ str(elapsed_time) + \" to deal with the page\")\n",
    "\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88c7197f",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(full_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "98b01ba7",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    " \n",
    "\n",
    "full_df = pd.DataFrame(full_data, columns=col_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a043d756",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "url = \"https://pford.info/refolddatabase/refoldingrecord\"\n",
    "r = requests.get(url)\n",
    "soup = BeautifulSoup(r.text, \"html.parser\")\n",
    "data_columns = [th.text.strip() for th in soup.find_all('th')]\n",
    "data_columns.append(\"url\")\n",
    "data = []\n",
    "for row in soup.find_all('tr'):\n",
    "    row_data = []\n",
    "    for cell in row.find_all(\"td\"):\n",
    "        link = cell.find(\"a\")\n",
    "        if link:\n",
    "            url = \"https://pford.info/refolddatabase/refoldingrecord/\" + link.get(\"href\")\n",
    "        row_data.append(cell.text)\n",
    "    row_data.append(url)\n",
    "    data.append(row_data)\n",
    "\n",
    "# remove the first entry as it is the headers\n",
    "data = data[1:]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "cf170948",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "while soup.find(\"a\", string=\"Next\"):\n",
    "    url = \"https://pford.info/refolddatabase/refoldingrecord\"+soup.find(\"a\", string=\"Next\").get(\"href\")\n",
    "    r = requests.get(url)\n",
    "    sub_data = []\n",
    "    soup = BeautifulSoup(r.text, \"html.parser\")\n",
    "    for row in soup.find_all('tr'):\n",
    "        row_data = []\n",
    "        for cell in row.find_all(\"td\"):\n",
    "            link = cell.find(\"a\")\n",
    "            if link:\n",
    "                url = \"https://pford.info/refolddatabase/refoldingrecord/\" + link.get(\"href\")\n",
    "            row_data.append(cell.text)\n",
    "        row_data.append(url)\n",
    "        sub_data.append(row_data)\n",
    "    \n",
    "\n",
    "    # remove the first entry as it is the headers\n",
    "    sub_data = sub_data[1:]\n",
    "\n",
    "    data.extend(sub_data)\n",
    "    break\n",
    "\n",
    "Monash_df = pd.DataFrame(data, columns=data_columns)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "388adc3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Protein'"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "soup.find_all(\"th\")[0].find(string=True, recursive=False).strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "c39174e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "r = requests.get(\"https://pford.info/refolddatabase/refoldingrecord/5/\")\n",
    "\n",
    "soup = BeautifulSoup(r.text, \"html.parser\")\n",
    "header = \"\"\n",
    "col_names = []\n",
    "data = []\n",
    "\n",
    "for th in soup.find_all(\"th\"):\n",
    "    if th.get(\"class\")[0] == \"detail_header\":\n",
    "        header = th.find(string=True, recursive=False).strip()\n",
    "    else:\n",
    "        col_names.append(header + \".\" + th.text.replace(\" \",\"_\"))\n",
    "\n",
    "\n",
    "for row in soup.find_all('tr'):\n",
    "        for cell in row.find_all(\"td\"):\n",
    "            data.append(cell.text)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "fd9841a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "57"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(col_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed3cc3ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "uniprot_id\n",
      "NaN       87\n",
      "Q90W39    36\n",
      "P08037    30\n",
      "P0A917    27\n",
      "P00698    16\n",
      "C3W5U5    14\n",
      "P09611    13\n",
      "P22629    12\n",
      "P09919    12\n",
      "P01579    12\n",
      "P08183    10\n",
      "P15692    10\n",
      "P98073     9\n",
      "P0C1Z0     9\n",
      "P22621     9\n",
      "Q6LEL2     9\n",
      "P45880     8\n",
      "P01241     8\n",
      "P01236     8\n",
      "P48061     8\n",
      "Name: count, dtype: int64\n",
      "total rows: 1875\n",
      "unique uniprot_id (non-null): 1083\n",
      "missing uniprot_id: 87\n",
      "uniprot_ids appearing more than once: 261\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# counts per uniprot_id (including NaN)\n",
    "counts = df['uniprot_id'].value_counts(dropna=False)\n",
    "print(counts.head(20))\n",
    "\n",
    "# summary stats\n",
    "total = len(df)\n",
    "unique = df['uniprot_id'].nunique(dropna=True)\n",
    "n_missing = df['uniprot_id'].isna().sum()\n",
    "n_duplicated_ids = (counts > 1).sum()\n",
    "\n",
    "print(f\"total rows: {total}\")\n",
    "print(f\"unique uniprot_id (non-null): {unique}\")\n",
    "print(f\"missing uniprot_id: {n_missing}\")\n",
    "print(f\"uniprot_ids appearing more than once: {n_duplicated_ids}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "072283ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pubmed_id                                                            20920547.0\n",
       "title                         Mutagenesis analyses explore residues responsi...\n",
       "abstract                      Trimucrotoxin (TmCT) is an Asn(6)-containing p...\n",
       "date                                                                   2011 Jan\n",
       "author                                 Tsai Inn-Ho,Wang Ying-Ming,Hseu Ming-Jhy\n",
       "journal                                                 Biochimie. 93(2) 277-85\n",
       "name                                                              Trimucrotoxin\n",
       "aaseq                         NLLQ FQKMIKIMTK KNAIPFYSSY GCYCGWGGQG KPKDATDR...\n",
       "comment                       without signal peptide, N-terminal His6 tag, N...\n",
       "uniprot_id                                                               Q90W39\n",
       "function                      Hydrolase, Myotoxin, Neurotoxin, Presynaptic n...\n",
       "domain                                                                      NaN\n",
       "ph                                                                          9.0\n",
       "temperature                                                                   4\n",
       "dilution                                                                      f\n",
       "dialysis                                                                      t\n",
       "column_filtration                                                             f\n",
       "column_binding                                                                f\n",
       "high_pressure                                                                 f\n",
       "other_method                                                                  f\n",
       "activity                                                                      t\n",
       "solubility                                                                    f\n",
       "non_aggregability                                                             f\n",
       "circular_dichroism                                                            t\n",
       "fluorescence_tryptophan                                                       f\n",
       "nuclear_magnetic_resonance                                                    f\n",
       "crystallization                                                               f\n",
       "structure_determination                                                       f\n",
       "Name: 1311, dtype: object"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "549100ee",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
